### ISSUE TITLE
JsonReadFeature: Introduce missing features for JSON5 compatibility

### ISSUE THREAD

**Issue by GrahamLea** [FasterXML/jackson-core#707]
The optional features in the JSON parser allow it be configured to _almost_ be a fully-compatible parser for JSON5.

There's just a few missing features, documented here:
https://stackoverflow.com/questions/68312227/can-the-jackson-parser-be-used-to-parse-json5

Summarised as:
- hexadecimal numbers
- trailing decimal points on numbers
- plus signs before numbers
- extra whitespace characters

I propose that additional Jackson features be added for each of these JSON5 features to allow Jackson to be used as a fully-compliant JSON5 parser.

(#612 was a similar but more specific request, though it seems that feature is already covered)

**Comment by cowtowncoder** [comment]
Sounds reasonable, as long as these are opt-in features (or, possibly, one JSON5 feature), to be added in `JsonReadFeature` and not defaults.
I probably won't have time to work on this in near future but as usual would be happy to help someone else has time for a PR or two!

**Comment by cowtowncoder** [comment]
So, thanks to @pjfanning 's work, 2 number related features needed here are implemented to go in 2.14!

**Comment by cowtowncoder** [comment]
Ok, so looks like we have 2 things missing:

* Hexadecimal numbers
* Extra whitespace 

Of these, hex numbers might be relatively easy to add (esp. for blocking parsers).
White space could be trickier to do efficiently but maybe not (esp. for "common" white space like tabs and other low-numeric code ones); also seems like less commonly needed feature.

**Comment by cowtowncoder** [comment]
If anyone is interested, filing a separate issue for, say, Hexadecimal number support; adding unit tests (failing, under `.../failing/`) would be a way to get things started.

**Comment by cowtowncoder** [comment]
Noting that if added, should be done for 3.x, not 2.x.

**Comment by seonwooj0810** [comment]
@cowtowncoder I'd like to take this on for 3.x, starting with hexadecimal number support
only, per your 2023-05-30 suggestion to "file a separate issue for, say, Hexadecimal
number support" and your 2025-11-17 confirmation that this should target 3.x.

Plan I'd like to validate before opening a PR:

**Scope (this PR)**: Hexadecimal integer support only — `0x1A`, `0X1A`, optionally with
sign (`-0x1A`, `+0x1A` when `ALLOW_LEADING_PLUS_SIGN_FOR_NUMBERS` is also enabled).
Float and exponent forms are out of scope.

**API**: New `JsonReadFeature.ALLOW_HEXADECIMAL_NUMBERS` (default `false`), added next
to the existing `ALLOW_LEADING_PLUS_SIGN_FOR_NUMBERS` / `ALLOW_LEADING_ZEROS_FOR_NUMBERS`
in the "number representations" block.

**Implementation surface (this PR)**: blocking parsers only, i.e.
- `ReaderBasedJsonParser` (char input)
- `UTF8StreamJsonParser` (byte input)

`UTF8DataInputJsonParser` and `NonBlockingUtf8JsonParserBase` would be a follow-up PR
(extra whitespace from #707 also follow-up). Happy to do those next if you'd prefer
them bundled, but starting narrow per your guidance.

**Test plan** (`NonStandardHexadecimalNumbers707Test` under `unittest/read/`,
modeled on `NonStandardLeadingPlusSign784Test`):
1. Basic positive: `0x0`, `0x1A`, `0xFF`, `0XAB`
2. With sign: `-0x1A`, and `+0x1A` only when `ALLOW_LEADING_PLUS_SIGN_FOR_NUMBERS` also enabled
3. Boundary: max int / max long / overflow → matches the existing int/long behavior
4. Regression: `0x1A` without feature → parse error with hint message
   ("enable `JsonReadFeature.ALLOW_HEXADECIMAL_NUMBERS` to allow")
5. Round-trip via `getString()` / `getNumberValue()`

**Two design choices I'd like your input on**:
- Does hex have to participate in `ALLOW_LEADING_ZEROS_FOR_NUMBERS` semantics
  (e.g., is `0x01A` allowed)? My read of JSON5 says yes, but happy to follow your
  preference.
- For the parse error message, should the hint reference both the feature name and
  the spec deviation, matching the existing `ALLOW_LEADING_PLUS_SIGN_FOR_NUMBERS`
  error wording?

I'll wait for your nod before coding.

**Comment by pjfanning** [comment]
@seonwooj0810 if a new feature is added but only added to some of the parsers - you are aware that we will then have people complaining that the feature is missing from the other parsers and that that will increase the maintenance overhead on the existing maintainers? Is there a reason to just target ReaderBasedJsonParser and UTF8StreamJsonParser and ignore UTF8DataInputJsonParser and the NonBlocking parsers?
Any implementation will need to minimise the performance impact on users who don't enable this new feature.

**Comment by seonwooj0810** [comment]
@pjfanning Fair point — I'll target all four parsers (`ReaderBasedJsonParser`, `UTF8StreamJsonParser`, `UTF8DataInputJsonParser`, `NonBlockingUtf8JsonParserBase`) in a single PR rather than splitting. My original plan to start narrow was about review size, not a technical constraint, and I agree partial coverage would create a worse maintenance situation than a slightly larger PR.

On performance for users who don't enable the feature: my plan is to gate hex parsing behind the existing first-character dispatch — the `ALLOW_HEXADECIMAL_NUMBERS` flag is only consulted after we've already seen a `0` followed by `x`/`X`, so the standard numeric hot path takes zero additional branches. I'll include a JMH run (or extend an existing benchmark, whichever you prefer) comparing baseline vs. feature-disabled to confirm no regression before requesting review.

One remaining design question for either of you: should hex participate in `ALLOW_LEADING_ZEROS_FOR_NUMBERS` (i.e., is `0x01A` valid, or only `0x1A`)? JSON5 allows it, but I'll follow whatever you prefer.

**Comment by pjfanning** [comment]
@seonwooj0810 I don't have a strong opinion on what gets supported. I would suggest whatever JSON5 allows in terms of hex numbers.

ALLOW_LEADING_ZEROS_FOR_NUMBERS is not really about hex - it's about decimal numbers.

I would favour ignoring ALLOW_LEADING_ZEROS_FOR_NUMBERS for hex values and that ALLOW_HEXADECIMAL_NUMBERS allows any valid hex number as per JSON5 spec.

**Comment by cowtowncoder** [comment]
+1 in general & for @pjfanning 's comment on completing support for all 4 backends.

Timing-wise I think this may need to go in 3.3 (hoping to release 3.1 soon), but you can target `3.x` branch either way and we'll see where this would land.

**Comment by cowtowncoder** [comment]
On

> Should hex participate in ALLOW_LEADING_ZEROS_FOR_NUMBERS (i.e., is 0x01A valid, or only 0x1A)? JSON5 allows it, but I'll follow whatever you prefer.

I think that if ALLOW_HEXADECIMAL_NUMBERS` allowed, then leading "0" as part of "0x" would need to be allowed (obviously), but not other leading zero cases (for which `ALLOW_LEADING_ZEROS_FOR_NUMBERS` need to be enabled).

**Comment by cowtowncoder** [comment]
This issue limited to just adding support for hexadecimal numbers -- if whitespace still needed, can file a follow-up issue.
